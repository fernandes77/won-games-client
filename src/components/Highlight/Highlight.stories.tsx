import type { Meta, StoryObj } from '@storybook/react'
import { Highlight } from './Highlight'

const meta: Meta<typeof Highlight> = {
  component: Highlight
}

export default meta
type Story = StoryObj<typeof Highlight>

const commonArgs = {
  title: 'Red Dead Redemption is back',
  subtitle: 'Come see Johnâ€™s new adventures',
  backgroundImage: '/img/red-dead-img.jpg',
  buttonLabel: 'Buy now',
  buttonLink: '/rdr2'
}

export const Primary: Story = {
  args: commonArgs,
  decorators: [
    (Story) => (
      <div style={{ maxWidth: '65rem' }}>
        <Story />
      </div>
    )
  ]
}

export const WithFloatImage: Story = {
  args: { ...commonArgs, floatImage: '/img/red-dead-float.png' },
  decorators: [
    (Story) => (
      <div style={{ maxWidth: '65rem' }}>
        <Story />
      </div>
    )
  ]
}
