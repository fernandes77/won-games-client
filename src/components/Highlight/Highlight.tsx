import { Button } from '@/components/Button/Button'
import { twMerge } from 'tailwind-merge'

export type HighlightProps = {
  title: string
  subtitle: string
  backgroundImage: string
  floatImage?: string
  alignment?: 'right' | 'left'
  buttonLabel: string
  buttonLink: string
  className?: string
}

export const Highlight = ({
  title,
  subtitle,
  backgroundImage,
  floatImage,
  alignment = 'right',
  buttonLabel,
  className,
  buttonLink
}: HighlightProps) => {
  const alignmentMap = {
    left: 'grid-areas-[content_floatimage] grid-cols-[2fr_1.3fr]',
    right: 'grid-areas-[floatimage_content] grid-cols-[1.3fr_2fr]'
  }
  const justifyFloatImageMap = { left: 'justify-self-end', right: '' }
  const textAlignmentMap = { left: 'text-left', right: 'text-right' }

  return (
    <div
      style={{ backgroundImage: `url(${backgroundImage})` }}
      className={twMerge(
        `highlight relative grid ${alignmentMap[alignment]} h-56 bg-center bg-cover grid-cols-[1.3fr_2fr] min-md:h-80
        after:absolute after:w-full after:h-full after:bg-[rgba(0, 0, 0, 0.6)]
        `,
        className
      )}
    >
      {!!floatImage && (
        <img
          className={`${justifyFloatImageMap[alignment]} grid-in-[floatimage]
          z-base max-h-56 max-w-full self-end min-md:max-h-80`}
          src={floatImage}
          alt={title}
        />
      )}
      <div
        className={`z-base p-xs grid-in-[content] min-md:self-end min-md:p-lg ${textAlignmentMap[alignment]}`}
      >
        <h2 className="text-lg font-bold text-white min-md:text-xxl">
          {title}
        </h2>
        <h3 className="text-sm font-light text-white mb-md min-md:text-lg">
          {subtitle}
        </h3>
        <Button as={{ type: 'link', props: { href: buttonLink } }}>
          {buttonLabel}
        </Button>
      </div>
    </div>
  )
}
