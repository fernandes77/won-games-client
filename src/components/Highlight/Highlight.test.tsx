import { render, screen } from '@/utils/tests/helpers'
import { Highlight } from './Highlight'

const props = {
  title: 'Heading 1',
  subtitle: 'Heading 2',
  backgroundImage: '/img/red-dead-img.jpg',
  buttonLabel: 'Buy now',
  buttonLink: '/games/rdr2'
}

describe('<Highlight />', () => {
  it('should render headings and button', () => {
    const { container } = render(<Highlight {...props} />)

    expect(
      screen.getByRole('heading', { name: /heading 1/i })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('heading', { name: /heading 2/i })
    ).toBeInTheDocument()

    expect(screen.getByRole('link', { name: /buy now/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render background image', () => {
    const { container } = render(<Highlight {...props} />)

    expect(container.firstChild).toHaveStyle({
      backgroundImage: `url(${props.backgroundImage})`
    })
  })

  it('should render float image', () => {
    render(<Highlight {...props} floatImage="/float-image.png" />)

    expect(screen.getByRole('img', { name: props.title })).toHaveAttribute(
      'src',
      '/float-image.png'
    )
  })

  it('should render align right by default', () => {
    const { container } = render(<Highlight {...props} />)

    expect(container.firstChild).toHaveClass('grid-areas-[floatimage_content]')
  })

  it('should render align left by default', () => {
    const { container } = render(<Highlight {...props} alignment="left" />)

    expect(container.firstChild).toHaveClass('grid-areas-[content_floatimage]')
  })
})
