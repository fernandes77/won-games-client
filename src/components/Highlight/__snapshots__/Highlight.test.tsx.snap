// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Highlight /> should render headings and button 1`] = `
<div
  class="highlight relative grid grid-areas-[floatimage_content] h-56 bg-center bg-cover grid-cols-[1.3fr_2fr] min-md:h-80 after:absolute after:w-full after:h-full after:bg-[rgba(0, 0, 0, 0.6)]"
  style="background-image: url(/img/red-dead-img.jpg);"
>
  <div
    class="z-base p-xs grid-in-[content] min-md:self-end min-md:p-lg text-right"
  >
    <h2
      class="text-lg font-bold text-white min-md:text-xxl"
    >
      Heading 1
    </h2>
    <h3
      class="text-sm font-light text-white mb-md min-md:text-lg"
    >
      Heading 2
    </h3>
    <a
      class="h-10 text-sm px-md py-xxs hover:bg-gradient-to-t hover:from-[#e35565] hover:via-[#d958a6] hover:to-[#d958a6] inline-flex items-center justify-center cursor-pointer bg-gradient-to-t from-[#ff5f5f] via-[#f062c0] to-[#f062c0] text-white rounded p-xxs [&>svg]:w-4"
      href="/games/rdr2"
    >
      <span
        class=""
      >
        Buy now
      </span>
    </a>
  </div>
</div>
`;
